cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/tx/WeakReferencePromotingTransactionContext.java

```diff
@@ -0,0 +1,150 @@
+import com.google.common.collect.Sets;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Collection;
+import javax.annotation.Nullable;
+
+/**
+ * This is an implementation of TransactionContext that maintains some of its transaction-awares as
+ * weak references, to allow them to be garbage-collected between transactions:
+ * <ul>
+ *   <li>When a new transaction is started, it promotes all weak references to strong references
+ *   (by de-referencing before adding them to the transaction);</li>
+ *   <li>When the transaction finishes (either through finish() or abort()), it discards all strong
+ *   references;</li>
+ *   <li>All other transaction-awares (not instances of {@link WeakReferenceDelegatingTransactionAware})
+ *   are passed through to the transaction as is.</li>
+ * </ul>
+ * This makes sure that after the transaction is finished, the weak-references can be garbage-collected. But
+ * for the lifetime of a transaction, they are protected from garbage collection to ensure that they can
+ * participate in commit or rollback.
+ *
+ * Note that because TransactionContext is not an interface, we must extend the existing implementation
+ * even though this implementation in fact delegates to another TransactionContext rather than extending one.
+ */
+public class WeakReferencePromotingTransactionContext extends TransactionContext {
+
+  private static final Logger LOG = LoggerFactory.getLogger(WeakReferencePromotingTransactionContext.class);
+
+  private final TransactionSystemClient txClient;
+  private final Collection<TransactionAware> txAwares;
+  private TransactionContext txContext;
+
+  public WeakReferencePromotingTransactionContext(TransactionSystemClient txClient) {
+    super(txClient);
+    this.txClient = txClient;
+    this.txAwares = Sets.newLinkedHashSet();
+  }
+
+  @Override
+  public boolean addTransactionAware(TransactionAware txAware) {
+    TransactionAware toAdd = txAware;
+    // promote weak references to the actual tx-aware before adding to actual tx-context
+    if (txAware instanceof WeakReferenceDelegatingTransactionAware) {
+      toAdd = ((WeakReferenceDelegatingTransactionAware) txAware).get();
+      if (toAdd == null) {
+        // this should never happen. It means that a tx-aware was already gc'ed before it was added here
+        LOG.warn("Attempt to add weak-referenced transaction-aware '{}' that has already been garbage-collected");
+        return false;
+      }
+    }
+    if (!txAwares.add(txAware)) {
+      return false; // it must already be in the actual tx-context
+    }
+    if (txContext != null && !txContext.addTransactionAware(toAdd)) {
+      // this must be a duplicate weak-reference to the same tx-aware, which should never happen
+      LOG.warn("Attempt to add weak-referenced transaction-aware '{}' that was already added " +
+                 "either directly or through a different weak reference");
+    }
+    return true;
```

Comment:
It is always true in that case, because the txAwares.add() in line 76 would detect that it is already there and return false.